ARROW-12512: [C++][Python][Dataset] Create CSV writer class and add Datasets support

lidavidm opened the pull request:
This refactors the CSV write support to expose an explicit CSV writer class, and adds Python bindings and Datasets support.

lidavidm commented:
@emkornfield would you be free to take a look (at least the CSV side, if not the Datasets side)? No rush of course.

jorisvandenbossche commented:
Does this also enable writing CSV with the dataset API in Python? (`write_dataset(..., format="csv")`

lidavidm commented:
@jorisvandenbossche I missed that: CsvFileFormat.make_write_options in Python needs to be updated as well.

nealrichardson commented:
There's probably a very small amount of wiring to propagate this up to the R `write_dataset()` function; up to you if you want to handle it here or make another JIRA for it.

lidavidm commented:
I threw in R support and found & fixed a bug with scanning CSV datasets with manually-specified names.

westonpace commented on python/pyarrow/_csv.pxd:
```diff
@@ -44,3 +44,11 @@ cdef class ReadOptions(_Weakrefable):
+
+cdef class WriteOptions(_Weakrefable):
+    cdef:
+        unique_ptr[CCSVWriteOptions] options
```
Why does this need to be a `unique_ptr`?  `CCSVWriteOptions` is pretty trivial.

lidavidm replied:
Mostly for consistency with the other options, and in case we add things to WriteOptions that would make it a non-standard layout type, in which case Cython will generate a lot of compiler warnings as it relies on sizeof.

westonpace commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1747,8 +1749,15 @@ cdef class CsvFileFormat(FileFormat):
 
-    def make_write_options(self):
-        raise NotImplemented("writing CSV datasets")
+    def make_write_options(self, WriteOptions options=None,
```
This is kind of confusing having a method named `make_write_options` that takes in an instance of `WriteOptions`.  Perhaps in C++ it wouldn't be so bad but for Python I think we might want something more understandable.

westonpace commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -3107,6 +3107,33 @@ def test_write_dataset_parquet(tempdir):
+    table = pa.table([
+        pa.array(range(20)), pa.array(np.random.randn(20)),
+        pa.array(np.repeat(['a', 'b'], 10))
+    ], names=["f1", "f2", "part"])
```
The column here is named `part` which makes me think it is going to be used for partitioning but that isn't actually done.  I'm not sure this is a problem as much as an observation.

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -403,34 +415,44 @@ class CSVConverter {
 
   static constexpr int64_t kColumnSizeGuess = 8;
+  io::OutputStream* sink_;
+  std::shared_ptr<io::OutputStream> owned_sink_;
```
shared_ptr seems strange in general for a OutputStream which seems for the most part should have only one owner.

lidavidm replied:
I agree it seems weird, but both the IPC and Parquet writers use shared_ptr for this.

pitrou replied:
Well, except that in Python any object can be shared, even if it's logically "owned" by something.

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -403,34 +415,44 @@ class CSVConverter {
-  ASSIGN_OR_RAISE(std::unique_ptr<CSVConverter> converter,
-                  CSVConverter::Make(table.schema(), pool));
-  return converter->WriteCSV(table, options, output);
+  ASSIGN_OR_RAISE(auto converter, MakeCSVWriter(output, table.schema(), options));
```
nit: should converter now be writer?  (same question below)

emkornfield commented on cpp/src/arrow/dataset/file_csv.h:
```diff
@@ -83,6 +82,35 @@ struct ARROW_DS_EXPORT CsvFragmentScanOptions : public FragmentScanOptions {
 
+class ARROW_DS_EXPORT CsvFileWriteOptions : public FileWriteOptions {
+ public:
+  /// Options passed to csv::MakeCSVWriter. use_threads is ignored
```
Is use_threads used elsewhere?  The way the code is structured threads could be used for the casts, so if it is important we might want to file a follow-up JIRA.

lidavidm replied:
I copied this from the equivalent IPC struct - it doesn't apply here since there's no such parameter of course.

emkornfield commented on python/pyarrow/_csv.pyx:
```diff
@@ -946,15 +951,44 @@ def write_csv(data, output_file, write_options=None,
         raise TypeError(f"Expected Table or RecordBatch, got '{type(data)}'")
+
+
+cdef class CsvWriter(_CRecordBatchWriter):
```
nit: as much as I appreciate Csv naming convention I think CSV is used everywhere else?

emkornfield commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1819,6 +1824,28 @@ cdef class CsvFragmentScanOptions(FragmentScanOptions):
                                         self.read_options)
 
 
+cdef class CsvFileWriteOptions(FileWriteOptions):
```
same nit on Csv vs CSV

lidavidm replied:
Unfortunately in the context of datasets (and only datasets) all other classes already use Csv.

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -282,65 +283,79 @@ Result<std::unique_ptr<ColumnPopulator>> MakePopulator(const Field& field, char
     return Status::OK();
   }
 
+  Status Close() override { return Status::OK(); }
```
other places I can think of (Buffer comes to mind), the way this is handled is passing a unique_ptr instead of a shared_ptr.

pitrou replied:
Right, we don't close either in other "writer" classes. This is more flexible, though of course in the general case not very useful.

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -282,65 +283,79 @@ Result<std::unique_ptr<ColumnPopulator>> MakePopulator(const Field& field, char
+        options_(options) {}
 
-  Status PrepareForContentsWrite(const WriteOptions& options, io::OutputStream* out) {
+  Status PrepareForContentsWrite() {
```
I think at one point I might have been using as a signal to see if header was written, but I can't really remember.  I agree it is strange and I don't have a strong justification for this pattern.  It might of been to avoid having to make a factory function for the private class.

pitrou replied:
Sounds good to me.

emkornfield commented:
Took a quick pass through, seems OK to me.  (didn't look at R stuff at all) and I agree with Weston's comments.

lidavidm commented:
Thanks for the reviews. I think I've addressed all feedback, minus the shared_ptr<OutputStream> - while this is weird, it is the pattern used by IPC and Parquet as well and I think we may was well be consistent across the formats. (Also, IPC exposes both the output-owning and output-borrowing APIs too, even though it expects the caller to close the stream in both cases.)

lidavidm commented:
Just to follow up - APIs like FileSystem return `shared_ptr<OutputStream>` so it would be very annoying to take `unique_ptr`. And we could just take only `OutputStream*` but IMO even if the caller is supposed to keep the pointer alive, it's safer to offer the option to take a `shared_ptr` by default to minimize mistakes.

lidavidm commented:
Rebased/fixed conflicts.

lidavidm commented:
Rebased/fixed conflicts again.

On the shared ptr vs raw pointer: I think the other issue is that we allow constructing a `shared_ptr<Writer>` at which point we should hold a `shared_ptr<OutputStream>` or `unique_ptr<OutputStream>`. But the filesystem interfaces only give a `shared_ptr<OutputStream>`, hence we can only take a shared_ptr here.

We could perhaps not allow you to get a `shared_ptr<Writer>` and only let you use `Writer` or `Writer*` and then only accept a `OutputStream*`.

emkornfield commented:
I don't think I had any objections before but i can rereview if you want.  The writer api stuff is unfortunate but doesn't need to be addressed here.

lidavidm commented:
Ah, ok, I was unsure how to deal with that part. I'd appreciate another look (though I don't think anything substantial has changed; I did have to rebase a few times) before I merge, though.

lidavidm commented:
Rebased and fixed conflicts here.

pitrou reviewed (approved):
LGTM.

lidavidm commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -403,36 +416,44 @@ class CSVConverter {
   std::shared_ptr<ResizableBuffer> data_buffer_;
   const std::shared_ptr<Schema> schema_;
-  MemoryPool* pool_;
+  WriteOptions options_;
```
I added the const.

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -282,65 +283,76 @@ Result<std::unique_ptr<ColumnPopulator>> MakePopulator(const Field& field, char
     }
-    return std::unique_ptr<CSVConverter>(
-        new CSVConverter(std::move(schema), std::move(populators), pool));
+    auto writer = std::shared_ptr<CSVWriterImpl>(new CSVWriterImpl(
```
nit: std::make_shared?

emkornfield commented on cpp/src/arrow/csv/writer.cc:
```diff
@@ -282,65 +283,76 @@ Result<std::unique_ptr<ColumnPopulator>> MakePopulator(const Field& field, char
-               std::vector<std::unique_ptr<ColumnPopulator>> populators, MemoryPool* pool)
-      : column_populators_(std::move(populators)),
-        offsets_(0, 0, ::arrow::stl::allocator<char*>(pool)),
+  CSVWriterImpl(io::OutputStream* sink, std::shared_ptr<io::OutputStream> owned_sink,
```
i guess this would need to be public to used std::make_shared.

emkornfield commented on cpp/src/arrow/csv/writer.h:
```diff
@@ -37,11 +40,34 @@ namespace csv {
+
+/// \brief Create a new CSV writer.
+///
+/// \param[in] sink output stream to write to
```
also note that ownership is not taken here?

emkornfield commented on python/pyarrow/_csv.pyx:
```diff
@@ -1010,15 +1015,44 @@ def write_csv(data, output_file, write_options=None,
+            CCSVWriteOptions c_write_options
+            CMemoryPool* c_memory_pool = maybe_unbox_memory_pool(memory_pool)
+        _get_write_options(write_options, &c_write_options)
+        c_write_options.io_context = CIOContext(c_memory_pool)
```
IOContext is new to me in general.  Should we be making new API's take that instead of Memory pool?

lidavidm replied:
I would say yes, since it wraps up a memory pool, thread pool, and cancellation token all in one.

emkornfield commented:
A few random nits, but the core C++ looks OK to me.

pitrou commented:
Thanks for the updates @lidavidm !
